chia/wallet/notification_manager.py

```diff
@@ -0,0 +1,97 @@
+from __future__ import annotations
+import dataclasses
+import logging
+
+from blspy import G2Element
+from typing import Any, Dict, List, Optional
+
+from chia.protocols.wallet_protocol import CoinState
+from chia.types.announcement import Announcement
+from chia.types.blockchain_format.coin import Coin
+from chia.types.blockchain_format.program import Program
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.types.coin_spend import CoinSpend
+from chia.types.spend_bundle import SpendBundle
+from chia.util.db_wrapper import DBWrapper2
+from chia.util.ints import uint64
+from chia.wallet.notification_store import Notification, NotificationStore
+from chia.wallet.transaction_record import TransactionRecord
+from chia.wallet.util.compute_memos import compute_memos_for_spend
+from chia.wallet.util.notifications import construct_notification
+
+
+class NotificationManager:
+    wallet_state_manager: Any
+    log: logging.Logger
+    notification_store: NotificationStore
+
+    @staticmethod
+    async def create(
+        wallet_state_manager: Any,
+        db_wrapper: DBWrapper2,
+        name: Optional[str] = None,
+    ) -> NotificationManager:
+        self = NotificationManager()
+        if name:
+            self.log = logging.getLogger(name)
+        else:
+            self.log = logging.getLogger(__name__)
+
+        self.wallet_state_manager = wallet_state_manager
+        self.notification_store = await NotificationStore.create(db_wrapper)
+        return self
+
+    async def potentially_add_new_notification(self, coin_state: CoinState, parent_spend: CoinSpend) -> bool:
+        if (
+            coin_state.spent_height is None
+            or not self.wallet_state_manager.wallet_node.config.get("accept_notifications", False)
+            or self.wallet_state_manager.wallet_node.config.get("required_notification_amount", 0)
+            > coin_state.coin.amount
+        ):
+            return False
+        else:
+            coin_name: bytes32 = coin_state.coin.name()
+            memos: Dict[bytes32, List[bytes]] = compute_memos_for_spend(parent_spend)
+            coin_memos: List[bytes] = memos.get(coin_name, [])
+            if (
+                len(coin_memos) == 2
+                and construct_notification(
+                    bytes32(coin_memos[0]), Program.to(coin_memos[1]).get_tree_hash(), uint64(coin_state.coin.amount)
+                ).get_tree_hash()
+                == coin_state.coin.puzzle_hash
+            ):
+                await self.notification_store.add_notification(
+                    Notification(
+                        coin_state.coin.name(),
+                        coin_memos[1],
+                        uint64(coin_state.coin.amount),
+                    )
+                )
+            return True
+
+    async def send_new_notification(
```

Comment:
how would sender know what's the minimal amount `target` is requiring?

Reply:
In my mind, they don't.  Sending a notification doesn't guarantee acknowledgement anyways, so you pay as much as you are willing to in order to get the message across.